### PR TITLE
Implement more fine-grained deconting of exported symbols

### PR DESCRIPTION
**Pull request by vrurg**
The problem with containerized exports has been narrowed down to EXPORT sub using hash for declaring exports where all values are getting wrapped into Scalars. To fix that method import on World got a named parameter :need-decont which signals that imported symbols need to be checked for their sigils. Unless a sigil is $ or & the symbol value is nqp::decont'ed.

This commit works for both rakudo/rakudo#2979 and rakudo/rakudo#3012.

### CODE REVIEW

**Comment by niner** [comment]
So the error is actually on the exporting side? Why not fix it there instead? What if someone actually wants to export a container?

**Comment by vrurg** [comment]
> So the error is actually on the exporting side? Why not fix it there instead? What if someone actually wants to export a container?

How can I fix it on the exporting side when `EXPORT` must return a hash where _everything_ is containerized? Moreover, if you re-export using `Foo::EXPORT::ALL::` or anything like that – you also get containerized symbols no matter what `is export` was applied to.

The only ultimate solution I see here is a `Map` type which would always bind its values. And then we would have to demand `EXPORT` to use it.

**Comment by niner** [comment]
EXPORt must return a Map, not necessarily a Hash. That's both documented: https://doc.perl6.org/language/modules#index-entry-sub_EXPORT and done so in the wild: https://github.com/niner/Inline-Perl5/blob/master/lib/Inline/Perl5.pm6#L941

**Comment by lizmat** [comment]
my %m is Map = a => 42; %m<a> = 666;  # Cannot change key 'a' in an immutable Map

**Comment by vrurg** [comment]
Ok, unless I have done something wrong, `Map` is really the solution for correct exporting. Moreover, it also correctly preserves re-exported symbols. I was worried if `FLATTENABLE_HASH` don't mangle with values, but it seemingly doesn't.

But this doesn't help with cases where people are returning `Hash`es. Perhaps it would make sense to make an additional check and do deconting only if a `Hash` is returned? To have DWIM in action because many seem to not understand the clear difference between the two.

**Comment by vrurg** [comment]
**UPDATE** Exporting via `Map` has a funny side effect. What is exported this way becomes immutable too. Basically, `Map` strips off containers on its own!

```
module Bar {
    our $bar = 42;
}

multi EXPORT {
    my %ex is Map = (
        '$bar' => $Bar::bar,
    );
    %ex;
}
```

```
use Bar;

say $bar.VAR.^name; # Int
$bar = 1; # Cannot assign to an immutable value
```

I don't see any alternative for now. And it's time to move to my gate anyway. ;)

File: src/Perl6/World.nqp [hunk 1]
```diff
@@ -1068,7 +1068,7 @@ class Perl6::World is HLL::World {
                 my $Map := self.find_symbol(['Map'], :setting-only);
                 if nqp::istype($result, $Map) {
                     my $storage := $result.hash.FLATTENABLE_HASH();
-                    self.import($/, $storage, $package_source_name, :need-decont);
+                    self.import($/, $storage, $package_source_name, :need-decont(!($result =:= $Map)));
```
Comment (niner) [inline src/Perl6/World.nqp]: Shouldn't that be $result.WHAT?

Comment (vrurg) [reply src/Perl6/World.nqp]: Sure it is. Thanks!

**Pull request merged**